### PR TITLE
Fix draggable legend disappearing when picking while use_blit=True

### PR DESCRIPTION
**Pull request by vittoboa**
Fix for the issue #28129. When using a draggable legend with use_blit=True and picking enabled, the bug caused the whole legend to disappear once a line in the legend was clicked.

### CODE REVIEW

**Comment by tacaswell** [comment]
I have confirmed that this fixes the problem...but I do not understand why.  I may be reading the logic wrong, but I think changing the `if` structure around has no effect on what code gets run under what conditions and the critical change is moving the `save_offset` above the draw/blit?

**Comment by tacaswell** [comment]
Sorry, I retract my comment, it does not seem to work reliably even with this change.

**Comment by vittoboa** [comment]
The critical change was to place the drawing logic inside the ``if self.got_artist`` part. This is because if someone clicks on a line in the legend it triggers two ``on_pick`` events: one for generally clicking on the legend and one for clicking on the line. In the previous code it would only draw once ``evt.artist == self.ref_artist`` was true, which was only true in the case of the legend, so the first event. Now, it handles the _legend_ event, which sets ``self.got_artist = True`` like before, then as ``self.got_artist`` is True it now also handles the _line_ event.

But in which scenarios did it not work for you? I have tried testing it again with the code example provided in the original bug description and can't spot any issue

**Comment by tacaswell** [comment]
Ah, I sorted out what happened.

I changed the `fig.canvas.draw()` in the example code to `fig.canvas.draw_idle()` and the flashing came back.

**Comment by vittoboa** [comment]
Ah okey I see, was that the only change? Because it seems to still work for me even when using ``fig.canvas.draw_idle()``

File: lib/matplotlib/offsetbox.py [hunk 1]
```diff
@@ -1506,6 +1506,9 @@ def on_release(self, event):
             self.finalize_offset()
             self.got_artist = False
             if self._use_blit:
+                self.ref_artist.draw(
```
Comment (tacaswell) [inline lib/matplotlib/offsetbox.py:1509]: You probably want to restore the background as well here or you will double-draw the legend (set it to have a colored and fraction alpha background to see the effect).

Comment (vittoboa) [reply lib/matplotlib/offsetbox.py:1510]: Ah yeah that's right, added the restoration!

**Comment by tacaswell** [comment]
>  ... the legend disappears when spamming on a line.

I drink too much ☕ !

**Comment by vittoboa** [comment]
Hi! Did any maintainer maybe get the chance to test this?

File: lib/matplotlib/offsetbox.py [hunk 2]
```diff
@@ -1486,25 +1486,30 @@ def on_motion(self, evt):
...
-            self.save_offset()
 
     def on_release(self, event):
         if self._check_still_parented() and self.got_artist:
             self.finalize_offset()
             self.got_artist = False
             if self._use_blit:
+                self.canvas.restore_region(self.background)
+                self.ref_artist.draw(
+                    self.ref_artist.figure._get_renderer())
```
Comment (QuLogic) [inline lib/matplotlib/offsetbox.py:1511]: This fits on a line:
```suggestion
                self.ref_artist.draw(self.ref_artist.figure._get_renderer())
```

**Review by QuLogic — APPROVED** [review]
I confirmed that this did fix the example.

File: lib/matplotlib/offsetbox.py (hunk 2 above)
Comment (vittoboa) [reply lib/matplotlib/offsetbox.py]: Updated with the suggested code :+1:

**Review by tacaswell — APPROVED** [review]
Also verified it worked!

**Comment by tacaswell** [comment]
Given that this is code we do not have unit tests for, and the docs failure is due to the config changing on main but not getting propogated to correctly (this is understood...we build the runner based on the config on the branch then manually create the merge commit to build the docs which leaves the requirements assuming a different base version of Python than the runner has) I'm going to go ahead and merge this before CI finishes.

**Pull request merged**